Libs/MRML/Logic/vtkMRMLColorLogic.cxx

```diff
@@ -1076,3 +1094,699 @@ vtkMRMLProceduralColorNode* vtkMRMLColorLogic::CopyProceduralNode(vtkMRMLColorNo
+      }
+    // now get the requested coding from the standard term
+    if (standardTerm.compare("CodeValue") == 0)
+      {
+      returnString = term.CodeValue;
+      }
+    else if (standardTerm.compare("CodeMeaning") == 0)
+      {
+      returnString = term.CodeMeaning;
+      }
+    else if (standardTerm.compare("CodingSchemeDesignator") == 0)
+      {
+      returnString = term.CodingSchemeDesignator;
+      }
+    }
+  return returnString;
+}
+
+//---------------------------------------------------------------------------
+bool vtkMRMLColorLogic::
+LookupLabelFromCategorization(ColorLabelCategorization& labelCat, int& label, const char *lutName)
+{
+  if (!this->TerminologyExists(lutName))
+    {
+    return false;
+    }
+
+  int labelFound = -1;
+
+  std::string inputTypeName, inputModifierName;
+
+  inputTypeName = labelCat.SegmentedPropertyType.CodeMeaning;
+  inputModifierName = labelCat.SegmentedPropertyTypeModifier.CodeMeaning;
+  std::transform(inputTypeName.begin(), inputTypeName.end(), inputTypeName.begin(), ::tolower);
+  std::transform(inputModifierName.begin(), inputModifierName.end(), inputModifierName.begin(), ::tolower);
+
+  ColorCategorizationMapType::const_iterator iter = this->colorCategorizationMaps[lutName].begin();
+  ColorCategorizationMapType::const_iterator iterEnd = this->colorCategorizationMaps[lutName].end();
+  for (;iter!=iterEnd;iter++)
+    {
+    ColorLabelCategorization mapCat = iter->second;
+
+    // fuzzy comparison rules:
+    //  -- property category is ignored
+    //  -- property type must be found in the mapping
+    //  -- if modifier is non-empty, must match as well
+    //  -- capitalization is ignored
+    //  -- look at the meaning, ignore codes and designators
+    std::string mapTypeName, mapModifierName;
+
+    mapTypeName = mapCat.SegmentedPropertyType.CodeMeaning;
+    mapModifierName = mapCat.SegmentedPropertyTypeModifier.CodeMeaning;
+    std::transform(mapTypeName.begin(), mapTypeName.end(), mapTypeName.begin(), ::tolower);
+    std::transform(mapModifierName.begin(), mapModifierName.end(), mapModifierName.begin(), ::tolower);
+
+    if (mapTypeName.find(inputTypeName) != std::string::npos)
+      {
+      // found match in category name
+      if (inputModifierName != "")
+        {
+        // modifier is not empty
+        if (mapModifierName.find(inputModifierName) != std::string::npos)
+          {
+          labelFound = iter->first;
+          break;
+          }
+        }
+      else
+        {
+        // modifier is empty, and category matches, assume this is the right
+        // term
+        labelFound = iter->first;
+        break;
+        }
+      }
+  }
+
+  if (labelFound != -1)
+    {
+    label = labelFound;
+    }
+  return (labelFound==-1) ? false: true;
+}
+
+//---------------------------------------------------------------------------
+bool vtkMRMLColorLogic::PrintCategorizationFromLabel(int label, const char *lutName)
```

Comment:
@fedorov It seems like this could use the << operator rather than be used by it.